File: dashboard/src/lib/api/events.ts
import { request } from "./client"

export interface EventResponse {
  id: string
  call_id: string
  agent_id: string
  tool_name: string
  verdict: string
  mode: string
  timestamp: string
  payload: Record<string, unknown> | null
  created_at: string
}

export interface EventFilters {
  agent_id?: string
  tool_name?: string
  verdict?: string
  mode?: string
  since?: string
  until?: string
  limit?: number
  offset?: number
}

export function listEvents(filters: EventFilters = {}) {
  const params = new URLSearchParams()
  for (const [key, value] of Object.entries(filters)) {
    if (value !== undefined) params.set(key, String(value))
  }
  const qs = params.toString()
  return request<EventResponse[]>(`/events${qs ? `?${qs}` : ""}`)
}

export interface HistogramBucketResponse {
  bucket_start: string
  allowed: number
  denied: number
  pending: number
  observed: number
}

export interface HistogramFilters {
  since: string
  until: string
  bucket_seconds: number
  agent_id?: string
  tool_name?: string
  verdict?: string
}

export function getEventHistogram(filters: HistogramFilters) {
  const params = new URLSearchParams()
  for (const [key, value] of Object.entries(filters)) {
    if (value !== undefined) params.set(key, String(value))
  }
  return request<HistogramBucketResponse[]>(`/events/histogram?${params}`)
}

File: src/edictum_server/schemas/events.py
"""Schemas for the event-ingestion endpoint."""

from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, Field


class EventPayload(BaseModel):
    """A single audit event emitted by an agent."""

    call_id: str = Field(..., description="Globally unique call identifier for dedup")
    agent_id: str = Field(..., description="Identifier of the reporting agent")
    tool_name: str = Field(..., description="Tool that was called")
    verdict: str = Field(..., description="Contract verdict (allow, deny, etc.)")
    mode: str = Field(..., description="Evaluation mode (enforce, observe, etc.)")
    timestamp: datetime = Field(..., description="When the event occurred (agent clock)")
    payload: dict[str, object] | None = Field(
        default=None, description="Optional extra data attached to the event"
    )


class ManifestContract(BaseModel):
    """A contract entry from an agent's local manifest."""

    id: str = Field(..., description="Contract identifier")
    type: str = Field(..., description="Contract type (pre, post, session)")
    tool: str | list[str] = Field(..., description="Tool name or list of tool patterns")
    mode: str = Field(..., description="Evaluation mode (enforce, observe)")


class AgentManifest(BaseModel):
    """Contract manifest pushed by Gate agents alongside events."""

    agent_id: str = Field(..., description="Agent identifier")
    policy_version: str = Field(..., description="Hash of the contract bundle")
    contracts: list[ManifestContract] = Field(default_factory=list)


class EventBatchRequest(BaseModel):
    """Batch of events sent by an agent in a single HTTP call."""

    events: list[EventPayload] = Field(
        ...,
        min_length=1,
        max_length=1000,
        description="One or more audit events",
    )
    agent_manifest: AgentManifest | None = Field(
        default=None, description="Optional contract manifest from Gate agents"
    )


class EventResponse(BaseModel):
    """A single audit event returned by the query endpoint."""

    id: str
    call_id: str
    agent_id: str
    tool_name: str
    verdict: str
    mode: str
    timestamp: datetime
    payload: dict[str, object] | None = None
    created_at: datetime


class EventIngestResponse(BaseModel):
    """Response after ingesting a batch of events."""

    accepted: int = Field(..., description="Number of newly stored events")
    duplicates: int = Field(..., description="Number of events skipped (already ingested)")


class HistogramBucketResponse(BaseModel):
    """A single time bucket with pre-classified event counts."""

    bucket_start: datetime
    allowed: int = 0
    denied: int = 0
    pending: int = 0
    observed: int = 0
